Unity-Technologies/Graphics
# [ShaderGraph] [bugfix 1257028] Get rid of target bitmask shadow state

### Purpose of this PR
Fix for https://fogbugz.unity3d.com/f/cases/1257028/

The core of the issue was that active Targets were stored in two locations: as a list of active targets (serialized) and as a bitmask that flagged the active targets in the list of valid targets.  Since the bitmask was not serialized, it does not update itself automatically with undo/redo.  In general having state stored in two locations and having to keep them in sync everywhere is fragile, so I changed it so we only store the list of active targets.

Merging this with a change Alex was working on to replace the bitmask UI with a reorderable list UI, we got rid of the bitmasks altogether, which also removes the limit of 32 Targets.

This change also stops the list from being reorderable; the active Target list is always sorted in alphabetical order, so there's no reason to give the user the impression that they can reorder it.  Originally the reordering was introduced to be able to place VFX Target first, as the VFX Target required it.   However, that requirement is no longer necessary, all of our existing Targets do not care about order, and we force-sort the active Target list on deserialization.  After discussion with the team, we decided it should not be user reorderable, so I removed the UI ability to do so, and also refactored a bit of the Active Target code to help ensure it stays in sorted order.  We determined that if future Targets needed ordering, we could create a source-code-level interface to control ordering, which would be better than a UI control.

This PR also makes a few improvements to the handling of unknown Targets.  Their name is now displayed as "ClassName (Unknown)", and a nice explanation text is added to their UI.  When an unknown Target is made inactive, we no longer immediately remove it from the potential Target list, so the user can add it back if they want.  Since we save the Target settings for potential Targets, if the user removes and adds back an Unknown Target, all of that Targets serialized settings will be unchanged.

![image](https://user-images.githubusercontent.com/28871759/94606504-6c71c780-024f-11eb-8b25-c433ec193ebc.png)

![image](https://user-images.githubusercontent.com/28871759/94605100-74306c80-024d-11eb-861a-a06ff988a4fc.png)

This change also ensures that removed Target data is retained (at least until the ShaderGraph window is closed or there is an Assembly Reload).  So whenever a Target is removed, the users can add it back and retain their old settings.  Checked that this works with Undo/Redo as well.

---
### Testing status

Added two automated tests to check expected Target invariants.

**Manual Tests**: What did you do?

Ran repro case, tested undo/redo/assembly reload with modifying target list.

Tested adding and removing SRP packages, to see that targets become marked as unknown properly, and their data is retained if they are removed and re-added, undo/redo, even while unknown.

Tested mixes of add/remove/undo/redo, to make sure all combinations retain Target settings.

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Notes for the reviewers you have assigned.
